File: javascript/121.best-time-to-buy-and-sell-stock.js
/*
 * @lc app=leetcode id=121 lang=javascript
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let buy = prices[0],
    profit = 0;
  for (let i = 1; i < prices.length; i++) {
    if (prices[i] - buy > profit) profit = prices[i] - buy;
    if (prices[i] < buy) buy = prices[i];
  }
  return profit;
};
// @lc code=end

File: rust/121.best-time-to-buy-and-sell-stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut buy, mut profit) = (prices[0], 0);
        for i in 1..prices.len() {
            buy = buy.min(prices[i]);
            profit = profit.max(prices[i] - buy);
        }
        return profit;
    }
}
// @lc code=end
